src/m3ninx/postings/roaring/bitmap_multi_readonly.go

```diff
@@ -0,0 +1,846 @@
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package roaring
+
+import (
+	"fmt"
+	"math/bits"
+	"sync"
+
+	"github.com/m3db/m3/src/m3ninx/postings"
+)
+
+// UnionReadOnly expects postings lists to all be read only.
+func UnionReadOnly(unions []postings.List) (postings.List, error) {
+	union := make([]readOnlyIterable, 0, len(unions))
+	for _, elem := range unions {
+		b, ok := elem.(readOnlyIterable)
+		if !ok {
+			return nil, ErrNotReadOnlyBitmap
+		}
+
+		union = append(union, b)
+	}
+
+	return newMultiBitmap(multiBitmapOptions{
+		op:    multiBitmapOpUnion,
+		union: union,
+	})
+}
+
+// IntersectAndNegateReadOnly expects postings lists to all be read only.
+func IntersectAndNegateReadOnly(
+	intersects []postings.List,
+	negates []postings.List,
+) (postings.List, error) {
+	intersect := make([]readOnlyIterable, 0, len(intersects))
```

Comment:
nit: prealloc a slice of `iterables := make([]readOnlyIterable, 0, len(intersects) + len(negates)0` and have both intersect and negate be taken from that, e.g. `intersect, negate := iterables[:len(intersects)], iterables[len(intersects):]`